test: refactor foundry test directory into unit, e2e, and integration tests

arthurgousset opened the pull request:
### Description

1. refactors foundry tests in `test-sol/` into 3 distinct directories:

    ```
    test-sol/unit/
    test-sol/e2e/
    test-sol/integration/
    ```

1. updates foundry tests on CI to leverage the new directory structure:

    ```
    # unit tests
    forge test --match-path "*test-sol/unit/*"
    
    # e2e tests
    forge test --match-path "*test-sol/e2e/*"
    
    # integration tests
    forge test --match-path "*test-sol/integration/*"
    ```

1. moves foundry scripts from `migrations_sol` to a new `scripts/foundry/` directory. We already have a directory for `scripts/truffle/` and `scripts/bash/`, so this makes it easier to find foundry-related bash scripts going forward.

### Other changes

1. splits the migration, integration, and e2e tests on CI into distinct steps.
    1. this saves time on CI since both the e2e and integration tests can use the same devchain. That means only a single step is required to generate the migrations and run a devchain. Previously, each step required generating migrations and running a devchain.
    2. this helps make the workflow more readable, and ensures error logs can be more easily associated with a specific step on CI, rather than a single bash script.
1. edits to remappings (renames a few, and updates the order in `foundry.toml` for better readability)
1. adds `repository` and `directory` properties in `package.json` of [`@celo/devchain-anvil`](https://www.npmjs.com/package/@celo/devchain-anvil). This ensures npmjs.org displays hyperlinks to this Github repository.
1. adds a `CONTRIBUTING.md` file to `migrations_sol`. This can be used to add more information regarding the anvil devchain, and migrations going forward.

    <img width="500" alt="image" src="https://github.com/celo-org/celo-monorepo/assets/46296830/58d40dd7-aa18-4d8d-ae5b-8285d60273dc">

1. adds a "How we work" section in the `README` file of [`@celo/devchain-anvil`](https://www.npmjs.com/package/@celo/devchain-anvil). This is helpful for 3rd-party developers that find the package on npmjs.org. The links help developers take action and help improve the anvil devchain.

    <img width="500" alt="image" src="https://github.com/celo-org/celo-monorepo/assets/46296830/7ccbfd05-dab5-4514-a4ba-8e354f5f9289">


### Tested

Yes, tested on CI and locally. All changes are added to CI workflows going forward.

### Related issues

- Fixes #11022 
- Related https://github.com/celo-org/celo-monorepo/pull/11020

### Backwards compatibility

Yes, everything is backwards.

### Documentation

No, we'll need to create better documentation for foundry-related workflows.

arthurgousset committed:
chore(foundry.toml): adds `@mento-core/...` remapping

Also moves existing mappings into groups for better readability

arthurgousset committed:
refactor(test-sol/common): moves files to `unit/common/`

Also updates imports respectively using remappings.

arthurgousset committed:
refactor(test-sol/identity): moves files to `unit/identity/`

Also updates imports using remappings where necessary.

arthurgousset committed:
refactor(test-sol/stability): moves files to `unit/stability/`

Also updates imports using remappings where necessary.

arthurgousset committed:
refactor(test-sol/voting): moves files to `unit/voting/`

Also updates imports using remappings where necessary.

arthurgousset committed:
refactor(test-sol/governance): moves files to `unit/governance/`

Also updates imports using remappings where necessary.

arthurgousset committed:
chore(test-sol): update missing remappings

With these changes all contracts compile and resolve correctly.

arthurgousset committed:
chore(workflows/protocol_tests): updates paths to `test-sol/unit`

arthurgousset committed:
chore(workflows/protocol_tests): adds integration and e2e tests

arthurgousset committed:
style(workflows/protocol_tests): refactors `forge test` for better readability

arthurgousset committed:
Merge branch 'master' into arthurgousset/test/refactor-test-sol

arthurgousset commented on .github/workflows/protocol_tests.yml:
```diff
@@ -97,8 +114,28 @@ jobs:
+        run: |
+          forge test -vvv \
+          --match-path "test-sol/e2e/*" \
+          --fork-url http://127.0.0.1:$ANVIL_PORT
```
Note-to-self: This line won't work until I merge this PR into my branch: 
- https://github.com/celo-org/celo-monorepo/pull/11020

I'm waiting for that PR to be reviewed.

arthurgousset replied:
That's done in https://github.com/celo-org/celo-monorepo/pull/11023/commits/77ea5cc29246f63878ba56729ac6e5d10d71a833

arthurgousset committed:
chore(migrations_sol): moves scripts to `scripts/foundry/

Moves all bash scripts from `migrations_sol` to `scripts/foundry/`, and updates paths where necessary. We already have a directory for `scripts/truffle/` and `scripts/bash/`, so this makes it easier to find foundry-related bash scripts.

arthurgousset commented on .github/workflows/protocol-devchain-anvil.yml:
```diff
@@ -87,7 +87,7 @@ jobs:
 
       - name: Generate migrations
-        run: ./migrations_sol/run_integration_tests_in_anvil.sh
+        run: ./scripts/foundry/run_integration_tests_in_anvil.sh
```
Note-to-self: Consider refactoring this like here for better readability and clearer error logs.

https://github.com/celo-org/celo-monorepo/blob/b0fb5f9f9a2299575b5ab4c674edcfc5762546b5/.github/workflows/protocol_tests.yml#L121-L129

arthurgousset replied:
This is done in https://github.com/celo-org/celo-monorepo/pull/11023/commits/c7a451a53d475e03155518fbfbc7f696219afcd0

arthurgousset committed:
Merge branch 'master' into arthurgousset/test/refactor-test-sol

arthurgousset committed:
test(governance/mock): move `MockGovernance` to `unit/` folder

arthurgousset committed:
refactor(foundry.toml): rename `migrations-sol/` remapping

arthurgousset committed:
refactor(workflows): refactor "run everything" step

Runs all tests in the `unit/` directory instead of all test files in the repo.
This makes sense from my perspective, because e2e tests (in the `e2e/` directory) and integration tests (in the `integration/` directory) require a connection to an anvil devchain serving at localhost.

The intent of this command is to ensure that no unit tests are forgotten, since unit tests are run explicitly by going through the directories above. But, the intention is not to run all tests in the repo generally.

arthurgousset committed:
style(workflows/protocol-devchain-anvil): splits migration into 2 steps

This helps the script becoming more readable, and ensure error logs are clearly associated with a workflow step rather than a single bash script.

arthurgousset committed:
chore(workflows): defines `ANVIL_PORT` env variable in workflow

Previously, the `$ANVIL_PORT` variable was exported and passed around as an env variable in a bash script.

But that required generating anvil migrations and running a devchain repeatedly.
Instead, the workflow does that once and different steps can access the devchain.

But, in that case the env variable is not passed around, so it has to be defined at the workflow level.

Source: https://docs.github.com/en/actions/learn-github-actions/variables

soloseng commented on .github/workflows/protocol_tests.yml:
```diff
@@ -94,11 +112,32 @@ jobs:
+          --match-path "test-sol/e2e/*" \
+          --fork-url http://127.0.0.1:${{ env.ANVIL_PORT }}
+
+      # TODO(Arthur): It does not seem necessary to kill Anvil since the GitHub workflow 
```
Is this TODO still needed?

Not necessary for running CI tests, but would be good to have when running locally. otherwise, the devchain stays running in the background. 

might be worth creating a wrapper bash script around `./scripts/foundry/create_and_migrate_anvil_devchain.sh` that keeps the devchain running in the foreground, it can easily be close.

A yarn cmd to run the foundry devchain would also be nice.

arthurgousset replied:
> Is this TODO still needed?

Great catch, thank you. The comment is not needed anymore. I removed it in https://github.com/celo-org/celo-monorepo/pull/11023/commits/61b6d7d64e09351635f5d6da509008d132e28a99.

arthurgousset replied:
> Not necessary for running CI tests, but would be good to have when running locally. otherwise, the devchain stays running in the background.
> 
> might be worth creating a wrapper bash script around ./scripts/foundry/create_and_migrate_anvil_devchain.sh that keeps the devchain running in the foreground, it can easily be close.
> 
> A yarn cmd to run the foundry devchain would also be nice.

Great point, I added two yarn commands to start and stop the devchain when working locally:

1. `yarn anvil-devchain:start` starts a new anvil devchain serving at localhost.

    For example:
    
    ```sh
    $ yarn anvil-devchain:start
    
    yarn run v1.22.22
    $ ./scripts/foundry/create_and_migrate_anvil_devchain.sh
    # ...
    Total elapsed time: 193 seconds
    ✨  Done in 193.09s.
    ```

1. `yarn anvil-devchain:stop` terminates any anvil nodes serving at localhost.


    For example:

    ```sh
    # in packages/protocol/ directory
    $ yarn anvil-devchain:stop
    
    yarn run v1.22.22
    $ ./scripts/foundry/stop_anvil.sh
    Connection to localhost port 8546 [tcp/*] succeeded!
    Killed Anvil
    ✨  Done in 0.11s.
    ```

arthurgousset replied:
Also added a `CONTRIBUTING.md` file in https://github.com/celo-org/celo-monorepo/pull/11023/commits/bd61a72cf8ee31d406a71985696ff40aa33d2da4

soloseng reviewed (approved):
Left one comment regarding devchain ease of use.

arthurgousset committed:
chore(workflows/protocol_tests): removes code comment

arthurgousset committed:
feat(scripts/foundry): adds `stop_anvil.sh`

arthurgousset committed:
refactor(scripts/foundry): use `stop_anvil.sh` to stop anvil

arthurgousset committed:
feat(protocol/package.json): adds `anvil-devchain:...` (`start` and `stop`)

arthurgousset committed:
refactor(scripts/foundry): use `stop_anvil.sh` to stop anvil

arthurgousset committed:
Merge branch 'master' into arthurgousset/test/refactor-test-sol

arthurgousset committed:
refactor(migrations_sol): update `@migrations-sol` remapping

Previous the remapping was called `@celo-migrations`.

arthurgousset committed:
docs(migrations_sol/README): renames README file

Also changes title to match NPM package name. This is a matter of personal preference. IMO this is a small improvement in readability for 3rd party users.

arthurgousset committed:
docs(scripts/foundry): adds more links to `package.json`

Adds `repository` and `directory` to `package.json`.
This ensures npmjs.org displays hyperlinks to the github repository.

arthurgousset committed:
docs(migrations_sol/CONTRIBUTING): adds MVP dev docs

We can complete this doc with additional information regarding the anvil devchain and migrations going forward.

arthurgousset committed:
docs(migrations_sol/README): adds "how we work" section

This is helpful for 3rd party developers that found this package on npmjs.org.
The links help developers take action and help improve the anvil devchain.

arthurgousset commented:
(for future reference)

Post merge and publishing workflow, I checked that the new URLs and new "How we work" section in the README file of [@celo/devchain-anvil](https://www.npmjs.com/package/@celo/devchain-anvil). Looks and works as expected:

<img width="1172" alt="image" src="https://github.com/celo-org/celo-monorepo/assets/46296830/0fa99c0d-de62-4052-a4d8-31a22b7c330d">

arthurgousset commented on packages/protocol/scripts/foundry/start_anvil.sh:
```diff
@@ -8,9 +8,9 @@ mkdir -p $ANVIL_FOLDER
 
 # create package.json
-echo "{\"name\": \"@celo/devchain-anvil\",\"version\": \"1.0.0\",\"description\": \"Anvil based devchain that contains core smart contracts of celo\",\"author\":\"Celo\",\"license\": \"LGPL-3.0\"}" > $TMP_FOLDER/package.json
+echo "{\"name\": \"@celo/devchain-anvil\",\"version\": \"1.0.0\",\"repository\": { \"url\": \"https://github.com/celo-org/celo-monorepo\", \"directory\": \"packages/protocol/migrations_sol\" },\"homepage\": \"https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/migrations_sol/README.md\",\"description\": \"Anvil based devchain that contains core smart contracts of celo\",\"auth ...
```
(for future reference) 
This is how I verified the JSON is valid

```sh
echo "{\"name\": \"@celo/devchain-anvil\",\"version\": \"1.0.0\",\"repository\": { \"url\": \"https://github.com/celo-org/celo-monorepo\", \"directory\": \"packages/protocol/migrations_sol\" },\"homepage\": \"https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/migrations_sol/README.md\",\"description\": \"Anvil based devchain that contains core smart contracts of celo\",\"author\":\"Celo\",\"license\": \"LGPL-3.0\"}" | jq
{
  "name": "@celo/devchain-anvil",
  "version": "1.0.0",
  "repository": {
    "url": "https://github.com/celo-org/celo-monorepo",
    "directory": "packages/protocol/migrations_sol"
  },
  "homepage": "https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/migrations_sol/README.md",
  "description": "Anvil based devchain that contains core smart contracts of celo",
  "author": "Celo",
  "license": "LGPL-3.0"
}
```
